src/Init/Lean/Elab/Quotation.lean

```diff
@@ -6,67 +6,85 @@ Authors: Sebastian Ullrich
 prelude
 import Init.Lean.Syntax
 import Init.Lean.Elab.ResolveName
+import Init.Lean.Elab.Term
 import Init.Lean.Parser -- TODO: remove after removing old elaborator
 
 namespace Lean
-namespace Elab
 
--- TODO: replace with quotations where possible
-private def const (n : Name) : Syntax :=
-Syntax.node `Lean.Parser.Term.id #[Syntax.ident none n.toString.toSubstring n [(n, [])]]
+/-- Reflect a runtime datum back to surface syntax (best-effort). -/
+class HasQuote (α : Type) :=
+(quote : α → Syntax)
 
-private def app (fn arg : Syntax) : Syntax :=
-Syntax.node `Lean.Parser.Term.app #[fn, arg]
+export HasQuote (quote)
 
-private def appN (fn : Syntax) (args : Array Syntax) : Syntax :=
-args.foldl app fn
+instance Syntax.HasQuote : HasQuote Syntax := ⟨id⟩
+instance String.HasQuote : HasQuote String := ⟨mkStxStrLit⟩
+instance Nat.HasQuote : HasQuote Nat := ⟨fun n => mkStxNumLit $ toString n⟩
 
--- TODO: unclear if the following functions could also be useful to other code
 private def quoteName : Name → Syntax
-| Name.anonymous => const `Lean.Name.anonymous
-| Name.str n s _ => appN (const `Lean.mkNameStr) #[quoteName n, mkStxStrLit s]
-| Name.num n i _ => appN (const `Lean.mkNameNum) #[quoteName n, mkStxNumLit $ toString i]
+| Name.anonymous => Unhygienic.run `(_root_.Lean.Name.anonymous)
+| Name.str n s _ => Unhygienic.run `(_root_.Lean.mkNameStr %%(quoteName n) %%(Lean.HasQuote.quote s))
```

Comment:
I am assuming you had to use `Lean.HasQuote.quote s` instead of just `quote s` because we are still using the old frontend, and `quoteSyntax` is not accessing the current namespace, open decls, and aliases in the old frontend. Correct?

Reply:
Correct